nova_vm/src/ecmascript/types/language/string/data.rs

```diff
@@ -2,13 +2,40 @@ use wtf8::{Wtf8, Wtf8Buf};
 
 use crate::heap::{CompactionLists, HeapMarkAndSweep, WorkQueues};
 
-#[derive(Debug, Clone, PartialEq, Eq)]
+#[derive(Debug, Clone)]
 pub struct StringHeapData {
-    // pub utf16_len: Option<u32>,
     pub(crate) data: StringBuffer,
+    pub(crate) mapping: IndexMapping,
 }
 
-#[derive(Debug, Clone, PartialEq, Eq)]
+impl PartialEq for StringHeapData {
+    fn eq(&self, other: &Self) -> bool {
+        // If both strings are static, we can compare their pointers directly.
+        if let (&StringBuffer::Static(self_str), &StringBuffer::Static(other_str)) =
+            (&self.data, &other.data)
+        {
+            if std::ptr::eq(self_str, other_str) {
+                return true;
+            }
+        }
+        self.as_str() == other.as_str()
+    }
+}
+impl Eq for StringHeapData {}
+
+#[derive(Debug, Clone)]
+pub(crate) enum IndexMapping {
```

Comment:
**suggestion:** Maybe make a `Unknown` variant (or just use `Option<IndexMapping>`) and make the mapping calculation lazy?

Reply:
I wonder how common it is to just create a string and pass it around without ever looking at its indices. We're not just talking indexing a character, but getting the length, slicing the string... Is that rare enough that making the mapping computation lazy is worth it?

Reply:
I'd wager that it's actually very common. eg. Object key names, "kind" or "type" discriminator strings, ... Fairly many strings seem like they're likely never indexed.

And some indexing-like operations can probably avoid indexing as well: includes and indexOf probably don't need to index since they're comparing like to like.